openingd/dualopend.c

```diff
@@ -95,8 +123,1303 @@ struct state {
+		u16 serial_id;
+
+		/* Reset their_complete to false every round,
+		 * they have to re-affirm every time  */
+		they_complete = false;
+
+		msg = sync_crypto_read(tmpctx, state->pps);
+		t = fromwire_peektype(msg);
+		switch (t) {
+		case WIRE_TX_ADD_INPUT: {
+			const u8 *tx_bytes, *redeemscript;
+			u16 max_witness_len;
+			u32 outnum, sequence;
+			size_t len;
+			struct bitcoin_tx *tx;
+			struct bitcoin_txid txid;
+			struct amount_sat amt;
+			struct tlv_tx_add_input_tlvs *add_tlvs =
+				tlv_tx_add_input_tlvs_new(tmpctx);
+
+			if (!fromwire_tx_add_input(tmpctx, msg, &cid,
+						   &serial_id,
+						   cast_const2(u8 **, &tx_bytes),
+						   &outnum, &sequence,
+						   &max_witness_len,
+						   cast_const2(u8 **, &redeemscript),
+						   add_tlvs))
+				peer_failed(state->pps, &state->channel_id,
+					    "Parsing tx_add_input %s",
+					    tal_hex(tmpctx, msg));
+
+			check_channel_id(state, &cid, &state->channel_id);
+			/*
+			 * BOLT-FIXME #2:
+			 * - if is the `initiator`:
+			 *   - MUST send even `serial_id`s
+			 * - MUST fail the transaction collaboration if:
+			 *   ...
+			 * - it receives a `serial_id` from the peer
+			 *   with the incorrect parity
+			 */
+			if (serial_id % 2 != 0)
+				peer_failed(state->pps, &state->channel_id,
+					    "Invalid serial_id rcvd. %u", serial_id);
+			/*
+			 * BOLT-FIXME #2:
+			 * - MUST fail the transaction collaboration if:
+			 *   ...
+			 *  - it recieves a duplicate `serial_id`
+			 */
+			if (psbt_has_serial_input(psbt, serial_id))
+				peer_failed(state->pps, &state->channel_id,
+					    "Duplicate serial_id rcvd. %u", serial_id);
+
+			/* Convert tx_bytes to a tx! */
+			len = tal_bytelen(tx_bytes);
+			tx = pull_bitcoin_tx(state, &tx_bytes, &len);
+			if (!tx)
+				peer_failed(state->pps, &state->channel_id,
+					    "Invalid tx sent.");
+
+			if (outnum >= tx->wtx->num_outputs)
+				peer_failed(state->pps, &state->channel_id,
+					    "Invalid tx outnum sent. %u", outnum);
+			/*
+			 * BOLT-FIXME #2:
+			 * - MUST fail the transaction collaboration if:
+			 *   ...
+			 *  - it receives an input that would create a
+			 *    malleable transaction id (e.g. pre-Segwit)
+			 */
+			if (!is_segwit_output(&tx->wtx->outputs[outnum], redeemscript))
+				peer_failed(state->pps, &state->channel_id,
+					    "Invalid tx sent. Not SegWit %s",
+					    type_to_string(tmpctx, struct bitcoin_tx, tx));
+
+			/*
+			 * BOLT- #2:
+			 *  - MUST NOT re-transmit inputs it has already received from the peer
+			 *  ...
+			 * - MUST fail the transaction collaboration if:
+			 *   ...
+			 *  - it receives a duplicate input to one it sent previously
+			 */
+			bitcoin_txid(tx, &txid);
+			if (psbt_has_input(psbt, &txid, outnum))
+				peer_failed(state->pps, &state->channel_id,
+					    "Unable to add input");
+
+			/*
+			 * FIXME: do this
+			 * BOLT- #2:
+			 * - MUST fail the transaction collaboration if:
+			 *   ...
+			 * - it receives an unconfirmed input
+			 */
```

Comment:
Why?

Reply:
This mattered when we weren't receiving the whole tx, as we looked up the tx on-chain to assure segwitness. Now that we're sending the whole tx, you're right, we can relax it.